kikuchipy/indexing/template_matching.py

```diff
@@ -0,0 +1,227 @@
+        Templates
+    keep_n : int, optional
+        Number of match results to keep for each pattern, by default 1
+    metric : Union[str, SimilarityMetric], optional
+        Similarity metric, by default "zncc"
+    compute : bool, optional
+        Whether to compute dask arrays before returning, by default True.
+        Is True if `n_slices` is given.
+    n_slices : int, optional
+        Number of templates slices to process sequentially, by default None
+
+    Returns
+    -------
+    template_indices, metric_result : Tuple[np.ndarray, np.ndarray]
+        Template indices and corresponding metric results with data shapes (ny*nx,keep_n).
+        Both arrays are sorted along keep_n axis according to metric used.
+    """
+    if n_slices is not None:
+        # Will (for now) dask compute regardless of compute param
+        return _template_match_slice_templates(
+            patterns,
+            templates,
+            keep_n=keep_n,
+            metric=metric,
+            n_slices=n_slices,
+        )
+
+    metric = SIMILARITY_METRICS.get(metric, metric)
+    if not isinstance(metric, SimilarityMetric):
+        raise ValueError(
+            f"{metric} must be either of {list(SIMILARITY_METRICS.keys())} "
+            "or an instance of SimilarityMetric. See make_similarity_metric."
+        )
+
+    # Expects signal data to be located on the two last axis for all scopes
+    sig_data_shape = patterns.shape[-2:]
+    t_sig_shape = templates.shape[-2:]
+    if sig_data_shape != t_sig_shape:
+        raise OSError(
+            f"The pattern {sig_data_shape} and template {t_sig_shape} "
+            "signal shapes are not identical."
+        )
+
+    if not metric._is_compatible(patterns.ndim, templates.ndim):
+        raise OSError(
+            f"The shape of patterns {patterns.shape} and templates {templates.shape} "
+            f"are not compatible with the scope {metric.scope} of {type(metric).__name__}"
+        )
+
+    similarities = metric(patterns, templates)
+    if not isinstance(similarities, da.Array):
+        similarities = da.from_array(similarities)
+
+    # ONE_TO_ONE
+    if similarities.shape == ():
+        s = np.array([similarities.compute()]) if compute else similarities
+        return np.array([0]), s
+
+    # If N is < keep_n => keep_n = N
+    keep_n = min(keep_n, len(templates))
+
+    match_result = (
+        similarities.argtopk(metric.sign * keep_n, axis=-1),
+        similarities.topk(metric.sign * keep_n, axis=-1),
+    )
+    if compute:
+        with ProgressBar():
+            match_result = da.compute(*match_result)
+
+    # Flattens the signal axis if not already flat
+    # This is foremost a design choice for returning standard outputs
+    if not metric.flat:
+        match_result = (
+            match_result[0].reshape(-1, keep_n),
+            match_result[1].reshape(-1, keep_n),
+        )
+    template_indices, metric_results = match_result
```

Comment:
That is right, only for corresponding with docs for clearity.

Reply:
I think the docstring `Returns` part should have two objects described each, instead of describing them together, and then simulated_indices and metric results returned as just `return simulated_indices, metric_results` without having to use a tuple or list?